cf-net/cf-net.c

```diff
@@ -0,0 +1,766 @@
+{
+    assert(opts);
+    char *file = args[0];
+    AgentConnection *conn = CFNetOpenConnection(hostname);// FIXME
+    if (conn == NULL)
+    {
+        return 1;
+    }
+    bool encrypt = true;
+    struct stat sb;
+    int r = cf_remote_stat(conn, encrypt, file, &sb, "file");
+    if (r!= 0)
+    {
+        printf("Could not stat: '%s'\n", file);
+    }
+    else
+    {
+        Log(LOG_LEVEL_INFO, "Detailed stat output:\n"
+                            "mode  = %jo, \tsize = %jd,\n"
+                            "uid   = %ju, \tgid = %ju,\n"
+                            "atime = %jd, \tmtime = %jd\n",
+            (uintmax_t) sb.st_mode,  (intmax_t)  sb.st_size,
+            (uintmax_t) sb.st_uid,   (uintmax_t) sb.st_gid,
+            (intmax_t)  sb.st_atime, (intmax_t)  sb.st_mtime);
+        CFNetStatPrint(file, sb.st_mode, hostname);
+    }
+
+    DisconnectServer(conn);
+    return 0;
+}
+
+static int invalid_command(const char *cmd)
+{
+    Log(LOG_LEVEL_ERR, "Invalid, see: 'cf-net help %s' for more info\n", cmd);
+    return -1;
+}
+
+
+static int CFNetGet(CFNetOptions *opts, const char* hostname, char **args)
+{
+    assert(opts);
+    assert(hostname);
+    assert(args);
+    AgentConnection *conn = CFNetOpenConnection(hostname);
+    if (conn == NULL)
+    {
+        return 1;
+    }
+    char *local_file = NULL;
+
+    // TODO: Propagate argv and argc from main()
+    int argc = 0;
+    while (args[argc] != NULL)
+    {
+        ++argc;
+    }
+
+    static struct option longopts[] = {
+         { "output",     required_argument,      NULL,           'o' },
+         { NULL,         0,                      NULL,           0   }
+    };
+    assert(opts != NULL);
+    if (argc <= 1)
+    {
+        return invalid_command("get");
+    }
+    extern int optind;
+    optind = 0;
+    extern char *optarg;
+    int c = 0;
+    int start_index = 1;
+    // TODO: Experiment with more user friendly leading - optstring
+    const char *optstr = "+o:"; // + means stop for non opt arg. :)
+    bool specified_path = false;
+    while ((c = getopt_long(argc, args, optstr, longopts, &start_index))
+            != -1)
+    {
+        switch (c)
+        {
+            case 'o':
+            {
+                local_file = xstrdup(optarg);
+                specified_path = true;
+                break;
+            }
+            case ':':
+            {
+                return invalid_command("get");
+                break;
+            }
+            case '?':
+            {
+                return invalid_command("get");
+                break;
+            }
+            default:
+            {
+                printf("Default optarg = '%s', c = '%c' = %i",
+                       optarg, c, (int)c);
+                break;
+            }
+        }
+    }
+    args = &(args[optind]);
+    argc -= optind;
+    const char *remote_file = args[0];
+
+    if(local_file == NULL)
+    {
+        char *last = strrchr(remote_file, '/');
```

Comment:
Use `basename()` instead, to support Windows backslashes.

And make sure you drop the `+ 1` in `xstrdup()` below when you do.